### PR TITLE
buiding server on linux

### PR DESCRIPTION
**Pull request by DCTech-lab**

### CODE REVIEW

**Comment by apohl79** [comment]
will take a look soonish. cheers.

File: Server/Source/AudioWorker.hpp [hunk 1]
```diff
@@ -103,8 +103,14 @@ class AudioWorker : public Thread, public LogTagDelegate {
     bool waitForData();
 
     template <typename T>
-    AudioBuffer<T>* getProcBuffer();
-
+    AudioBuffer<T>* getProcBuffer()
+    {
+ 		if constexpr (std::is_same<T, float>::value)
```
Comment (apohl79) [inline Server/Source/AudioWorker.hpp:108]: The codebase is on C++14 currently.

**Comment by apohl79** [comment]
Do you intend to add more to the PR? It is pretty much incomplete (screen capturing, keyboard & mouse support etc) and causes new compiler warnings. If this is really just to make it compile and use the the generic editor for plugins hosted on linux, you should
- resolve the compiler warnings (I have see some sign-conversion and c++17 warnings)
- don't build the server per default on linux
- run the code formatter (package/format.sh)

thanks for the PR!

File: Server/Source/AudioWorker.hpp (hunk 1 above)
Comment (DCTech-lab) [reply Server/Source/AudioWorker.hpp]: Thanks for reviewing the code. Sorry for the late reply, I am not much familiar with the process.  
My intention was to make the code compilable first, so that the other issues on linux can be fixed. That was the first step of the changes. The dummy functionality of CPU load and other things will be fixed later. But  found that basic functionality of audio flowing, plugin menu etc is working fine. 

About the review points:

FIXED    "don't build the server per default on linux."  (removed the part which trats server build on linux specially)  
FIXED    "run the code formatter (package/format.sh)" (ran the script and committed the changes)

    "resolve the compiler warnings (I have seen some sign-conversion and c++17 warnings)"
The compiler error in AudioWorker.hpp was because the issue in C++14  in this special case of template specialization. The fix i got on  StackOverflow and the same mentioned that the issue resolved in C++17. I will try to fix other warnings about unused variables and conversions etc.

Since I dont have proper setup, request you to please download the binary of server in release section of the branch and test it. 

Audio is fine. plugin list (dummy) for client menu was also fine. Test the screen capturing, vst plugin loading etc. May be you need to change the settings manually in cfg files as it doesn't show the tray icon . Priority of development  will be for scanning and loading plugins.  

Thanks
Durgesh Kumar

**Comment by DC-codes** [comment]
Hi  Apohl,

For native screen capturing, can't we use generic JUCE functions like createComponentSnapshot, as in below code: 
        auto editor = plugin->getActiveEditor();
	auto editorRectangle = editor->getBounds();
	auto image = editor->createComponentSnapshot(editorRectangle, true);
	JPEGImageFormat jpegFormat = JPEGImageFormat();
	jpegFormat.writeImageToStream(image, socket2Client);

Here we may not have fine control over brightness etc, but that is simple and plateform independent. Actually for linux part i thought to add in that way.

Also we can use generic JUCE key press events instead of plateform specific native code
 

Thanks,

**Comment by apohl79** [comment]
> Hi Apohl,
> 
> For native screen capturing, can't we use generic JUCE functions like createComponentSnapshot, as in below code: auto editor = plugin->getActiveEditor(); auto editorRectangle = editor->getBounds(); auto image = editor->createComponentSnapshot(editorRectangle, true); JPEGImageFormat jpegFormat = JPEGImageFormat(); jpegFormat.writeImageToStream(image, socket2Client);
> 
> Here we may not have fine control over brightness etc, but that is simple and plateform independent. Actually for linux part i thought to add in that way.
> 
> Also we can use generic JUCE key press events instead of plateform specific native code
> 
> Thanks,

Hey Durgesh,

this was my very first approach, but the problem is that you can't capture context menus using this approach.

Best,
Andreas

**Comment by DCTech-lab** [comment]
Fixed the warnings and did a rough implementation of screen capture and mouse / key handling. The changes in server.hpp is to remove conversion warning and was not due to my changes.

Could not test the feature as the plugin (chowmatrix from github) window doesnt show gui and later size becomes zero (though show plugin window button on from window client on reaper works). Needs fixing the plugin window issue first to test the feature. Code compiles and runs fine on windows and linux (xubuntu)

File: Server/Source/AudioWorker.hpp [hunk 2]
```diff
@@ -103,18 +103,10 @@ class AudioWorker : public Thread, public LogTagDelegate {
     bool waitForData();
 
     template <typename T>
-    AudioBuffer<T>* getProcBuffer();
-
-    template <>
-    AudioBuffer<float>* getProcBuffer() {
+    AudioBuffer<T>* getProcBuffer() {
```
Comment (apohl79) [inline Server/Source/AudioWorker.hpp:106]: Why was that change needed?

**Comment by apohl79** [comment]
Sorry for the delay. I just had one question, but I'm going to merge this soon. Thanks for this!

File: Server/Source/AudioWorker.hpp (hunk 2 above)
Comment (DCTech-lab) [reply Server/Source/AudioWorker.hpp:106]: because without that it was giving compilation error on Linux.
both the definitions were specializations and in class and MSVC accepts the previous code but gcc on linux is strict. 
we can proceed with specialization definition out of class  definition, but that in header file , gives multiple definition error at linking time.
I declared one as general template definition and moved other specialization in cpp file. 
Still only two definitions (for float and double) will be generated. Trying with any other type will give compiler error as default float*  type buffer return value,  will not match with intended type and require another explicit specialization.
